### PR TITLE
Paper: Keeping your Jupyter notebook code quality bar high (and production ready) with Ploomber

### PR DESCRIPTION
**Pull request by idomic**
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.

*Editor*: @mepa

*Reviewers*: @kajayi

### CODE REVIEW

**Comment by scoobies** [comment]
Hi!, I'm @scoobies, a friendly bot.

Type ```@scoobies help``` to discover how I can help you.

**Comment by deniederhut** [comment]
@idomic is this a full paper that you would like to have included in the proceedings? Or a poster that you would like to have archived?

**Comment by idomic** [comment]
@deniederhut this is the paper to include, it's sort of a self guided tutorial on achieving best practices. Is there anything I need to change?

**Comment by deniederhut** [comment]
Yup you will need to follow the instructions here for submitting papers to the proceedings : https://github.com/scipy-conference/scipy_proceedings#instructions-for-authors

We're expecting a new directory with your name in the `papers` directory, which should have a .rst file there and any supporting files for tables or images. I see a lot of other stuff, like notebooks and idea project files, which we can't publish. You can see an example of how to structure your submission here: https://github.com/scipy-conference/scipy_proceedings/tree/2022/papers/00_bibderwalt

**Comment by deniederhut** [comment]
@scoobies mark paper

**Comment by deniederhut** [comment]
@scoobies build paper

**Comment by scoobies** [comment]
building paper...

**Comment by scoobies** [comment]
{"data"=>{"info"=>"Build for paper idomic-2022 scheduled.  Note that builds are only executed if the current build attempt is more than 5 minutes old."}, "status"=>"success"}

**Comment by deniederhut** [comment]
It looks like this paper is only two pages long, and that there are no references -- is that right?

**Comment by deniederhut** [comment]
@idomic just following up -- have you had a chance to clean up the extra files in your submission? Thanks!

**Comment by deniederhut** [comment]
@scoobies mark needs review

**Comment by deniederhut** [comment]
@scoobies assign @mepa as editor

**Comment by scoobies** [comment]
Assigned! @mepa is now the editor

**Comment by idomic** [comment]
@deniederhut On it!

**Comment by deniederhut** [comment]
Thanks!

**Comment by govarguz** [comment]
Dear all, I agree on reviewing this paper!

**Comment by idomic** [comment]
> It looks like this paper is only two pages long, and that there are no references -- is that right?

Yes that's correct it's a self explanatory workshop, there are references in the end (section 9).
Also there's a jupyter notebook that users can leverage and run locally.

Also removed the .idea file, and followed the instructions (the .rst and the folder structure was there already).

**Comment by deniederhut** [comment]
Thanks Horacio!

**Comment by deniederhut** [comment]
@scoobies add @govarguz to reviewers

**Comment by scoobies** [comment]
@govarguz added to the reviewers list!

**Comment by deniederhut** [comment]
> Yes that's correct it's a self explanatory workshop, there are references in the end (section 9).
Also there's a jupyter notebook that users can leverage and run locally.

The only output of the publication system is a PDF of your paper and html page with the abstract that we post online -- the jupyter notebooks and anything else will be ignored. Please remove any materials that are not needed to compile the paper.

**Comment by deniederhut** [comment]
We've moved the deadline for first reviews back one week, to June 16th, to give us a little more time to find reviewers for all the papers this year. You can see the updated full timeline at https://github.com/scipy-conference/scipy_proceedings#timeline-for-2022.

**Comment by deniederhut** [comment]
@scoobies add @kajayi to reviewers

**Comment by scoobies** [comment]
@kajayi added to the reviewers list!

**Comment by deniederhut** [comment]
@idomic have you had a chance to clean out the extra files?

**Comment by idomic** [comment]
Yes I'm adding a few references and diagrams since it won't be the
interactive notebook I thought. I'll push a new rst for review mid week.

**Comment by idomic** [comment]
@deniederhut 

* I've started building the paper again, having some difficulties on mac, it's showing me some missing latex dependencies.
* I'm now at a point where I don't have this error but for some reason the paper isn't generated, it's outputting building and no paper.pdf.

Does it work if I'm saving the notebook as a pdf and commit it?
Also is there a way to keep the notebook so it's more interactive? Even if I'm adding images and diagrams it's way better if users can run it.

**Comment by deniederhut** [comment]
> Also is there a way to keep the notebook so it's more interactive?

We aren't really set up for that, unfortunately. The proceedings is published as a static PDF. Maybe you could add a link in the paper to a binder instance or something?

> Does it work if I'm saving the notebook as a pdf and commit it?

Sorry -- we'll need the source file to compile the proceedings.

> it's outputting building and no paper.pdf.

Hard to be sure what's happening there without more info. Is there output you can share? Or could you push your recent changes to this branch so I can pull them down and debug locally?

**Comment by idomic** [comment]
A binder link is a great idea!

On the changes, I'm trying to build the existing `.rst` even before the changes, the output of `./make_paper.sh papers/ido_michael/` is `Building: ido_michael` and that's it, when I'm looking for the final .pdf it's not there

**Comment by idomic** [comment]
Alright, I found where it outputs the .pdf, it's just in a different directory than what I was expecting. I'm adding the missing links and diagrams.

**Comment by idomic** [comment]
@deniederhut I've added the rest of the edits, the paper should be 4 pages long with more diagrams and links. Let me know if anything's missing.

* Should I just tag the reviewers instead?

**Comment by deniederhut** [comment]
Glad it was an easy fix! @govarguz and @kajayi -- friendly reminder that initial reviews are due tomorrow

File: papers/ido_michael/index.rst [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+   cd playground
+   ploomber plot
+
+.. image:: images/new-task-attached.png
+
+
+Ploomber now recognizes our dependency declaration!
+
+Open ``playground/tasks/gradient-boosting-regressor.py`` as a notebook by right-clicking on it and then ``Open With`` -> ``Notebook`` and add the following code:
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (kajayi) [inline papers/ido_michael/index.rst]: Consider putting this into a separate Markdown cell

File: papers/ido_michael/index.rst [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+   upstream = ['train-test-split']
+
+Let's generate the plot again:
+
+.. code:: sh
+
+   cd playground
+   ploomber plot
+
+.. image:: images/new-task-attached.png
```
Comment (kajayi) [inline papers/ido_michael/index.rst]: Needs a caption

File: papers/ido_michael/index.rst [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+
+At the top of the notebook, you’ll see the following:
+
+.. code:: python
+
+   upstream = None
+
+This special variable indicates which tasks should execute before the
+notebook we're currently working on. In this case, we want to get
+training data so we can train our new model so we change the
```
Comment (kajayi) [inline papers/ido_michael/index.rst:324]:
```suggestion
training data so we can train our new model, so we change the
```

File: papers/ido_michael/index.rst [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+.. code:: sh
+
+   cd playground
+   ploomber plot
+
+.. image:: images/new-task.png
+
+
+You can see that Ploomber recognizes the new file, but it does not have
+any dependency, so let’s tell Ploomber that it should execute after
```
Comment (kajayi) [inline papers/ido_michael/index.rst:298]:
```suggestion
any dependencies, so let’s tell Ploomber that it should execute after
```

File: papers/ido_michael/index.rst [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+
+
+Let's see how the plot looks now:
+
+.. code:: sh
+
+   cd playground
+   ploomber plot
+
+.. image:: images/new-task.png
```
Comment (kajayi) [inline papers/ido_michael/index.rst]: Needs caption

File: papers/ido_michael/index.rst [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+
+7. Adding a new task
+--------------------
+
+Let’s say we want to train another model and decide to try `Gradient
+Boosting
+Regressor <https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingRegressor.html#sklearn.ensemble.GradientBoostingRegressor>`__.
+First, we modify the ``pipeline.yaml`` file and add a new task:
+
+Open ``playground/pipeline.yaml`` and add the following lines at the end
```
Comment (kajayi) [inline papers/ido_michael/index.rst:256]:
```suggestion
First, we modify the ``pipeline.yaml`` file and add a new task:

Open ``playground/pipeline.yaml`` and add the following lines at the end:
```

File: papers/ido_michael/index.rst [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+The arrows in the diagram represent input/output dependencies, hence,
+determine execution order. For example, the first task (``load``) loads
+some data, then ``clean`` uses such data as input and process it, then
+``train-test-split`` splits our dataset in training and test, finally,
+we use those datasets to train a linear regression and a random forest
+regressor.
+
+Soorgeon extracted and declared this dependencies for us, but if we want
+to modify the existing pipeline, we need to declare such dependencies.
+Let’s see how.
```
Comment (kajayi) [inline papers/ido_michael/index.rst:243]:
```suggestion
Soorgeon extracted and declared the dependencies for us, but if we want
to modify the existing pipeline, we need to declare such dependencies manually.
Let’s see how.
```

File: papers/ido_michael/index.rst [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+
+.. image:: images/executed-pipeline.png
+
+
+The arrows in the diagram represent input/output dependencies, hence,
+determine execution order. For example, the first task (``load``) loads
+some data, then ``clean`` uses such data as input and process it, then
+``train-test-split`` splits our dataset in training and test, finally,
+we use those datasets to train a linear regression and a random forest
+regressor.
```
Comment (kajayi) [inline papers/ido_michael/index.rst]:
```suggestion
The arrows in the diagram represent input/output dependencies and depict the 
execution order. For example, the first task (``load``) loads
some data, then ``clean`` uses such data as input and processes it, then
``train-test-split`` splits our dataset into training and test sets. Finally,
we use those datasets to train a linear regression and a random forest
regressor.
```

File: papers/ido_michael/index.rst [hunk 9]
```diff
@@ -0,0 +1,403 @@
...
+6. Maintaining the pipeline
+---------------------------
+
+Let’s look again at our pipeline plot:
+
+.. code:: python
+
+   Image('playground/pipeline.png')
+
+.. image:: images/executed-pipeline.png
```
Comment (kajayi) [inline papers/ido_michael/index.rst]: Needs caption

File: papers/ido_michael/index.rst [hunk 10]
```diff
@@ -0,0 +1,403 @@
...
+are crucial to us before training the model.
+
+6. Maintaining the pipeline
+---------------------------
+
+Let’s look again at our pipeline plot:
+
+.. code:: python
+
+   Image('playground/pipeline.png')
```
Comment (kajayi) [inline papers/ido_michael/index.rst:228]: Delete?

File: papers/ido_michael/index.rst [hunk 11]
```diff
@@ -0,0 +1,403 @@
...
+    # Validating output type
+    assert 'pkl' in upstream['train-test-split']['X_test']
+
+Adding these snippets will allow us to validate that the data we're looking for exists
+as well as the quality we expect. For instance, in the first test we're checking
+there are no missing rows, and that the data sample we have are for houses up to 100
+years old.
+
+In the second snippet, we're checking that there are train and test inputs which
+are crucial to us before training the model.
```
Comment (kajayi) [inline papers/ido_michael/index.rst]:
```suggestion
are crucial for training the model.
```

File: papers/ido_michael/index.rst [hunk 12]
```diff
@@ -0,0 +1,403 @@
...
+    assert Path(upstream['train-test-split']['X_test']).exists()
+
+    # Check task train input exists
+    assert Path(upstream['train-test-split']['y_train']).exists()
+
+    # Validating output type
+    assert 'pkl' in upstream['train-test-split']['X_test']
+
+Adding these snippets will allow us to validate that the data we're looking for exists
+as well as the quality we expect. For instance, in the first test we're checking
```
Comment (kajayi) [inline papers/ido_michael/index.rst]:
```suggestion
and has the quality we expect. For instance, in the first test we're checking
```

File: papers/ido_michael/index.rst [hunk 13]
```diff
@@ -0,0 +1,403 @@
...
+.. image:: images/build-output.png
+
+Navigate to ``playground/output/`` and you’ll see all the outputs: the
+executed notebooks, data files and trained model.
+
+.. code:: sh
+
+   ls playground/output
+
+.. image:: images/post-build-artifacts.png
```
Comment (kajayi) [inline papers/ido_michael/index.rst]: Needs a caption

File: papers/ido_michael/index.rst [hunk 14]
```diff
@@ -0,0 +1,403 @@
...
+   cd playground
+   ploomber build
+
+We can see which are the tasks that ran during this command, how long did it take,
+and what's the partial percentage of the overall pipeline execution runtime.
+
+.. image:: images/build-output.png
+
+Navigate to ``playground/output/`` and you’ll see all the outputs: the
+executed notebooks, data files and trained model.
```
Comment (kajayi) [inline papers/ido_michael/index.rst:165]:
```suggestion
Navigate to ``playground/output/`` and you’ll see all the outputs&mdash;the
executed notebooks, data files and trained model:
```

File: papers/ido_michael/index.rst [hunk 15]
```diff
@@ -0,0 +1,403 @@
...
+
+.. code:: sh
+
+   cd playground
+   ploomber build
+
+We can see which are the tasks that ran during this command, how long did it take,
+and what's the partial percentage of the overall pipeline execution runtime.
+
+.. image:: images/build-output.png
```
Comment (kajayi) [inline papers/ido_michael/index.rst]: Needs a caption

File: papers/ido_michael/index.rst [hunk 16]
```diff
@@ -0,0 +1,403 @@
...
+
+Let’s build the pipeline (this will take ~30 seconds):
+
+.. code:: sh
+
+   cd playground
+   ploomber build
+
+We can see which are the tasks that ran during this command, how long did it take,
+and what's the partial percentage of the overall pipeline execution runtime.
```
Comment (kajayi) [inline papers/ido_michael/index.rst]:
```suggestion
We can see which tasks ran during this command, how long they took to execute,
and the contributions of each task to the overall pipeline execution runtime.
```

File: papers/ido_michael/index.rst [hunk 17]
```diff
@@ -0,0 +1,403 @@
...
+of each other!).
+
+We can also get a summary of the pipeline with ``ploomber status``:
+
+.. code:: sh
+
+   cd playground
+   ploomber status
+
+.. image:: images/status-output.png
```
Comment (kajayi) [inline papers/ido_michael/index.rst]: Needs a caption

File: papers/ido_michael/index.rst [hunk 18]
```diff
@@ -0,0 +1,403 @@
...
+task was executed or failed.
+
+
+.. image:: images/pipeline-step-2.png
+
+
+Soorgeon correctly identified the *stages* in our original ``nb.ipynb``
+notebook. It even detected that the last two tasks
+(``linear-regression``, and ``random-forest-regressor`` are independent
+of each other!).
```
Comment (kajayi) [inline papers/ido_michael/index.rst]:
```suggestion
notebook. It even detected that the last two tasks
(``linear-regression`` and ``random-forest-regressor``) are independent
of each other!
```

File: papers/ido_michael/index.rst [hunk 19]
```diff
@@ -0,0 +1,403 @@
...
+.. code:: sh
+
+   ploomber plot
+
+This command will generate the plot below for us, that way we can stay up to date
+with changes that are happening in our pipeline and get an immidate status of what
+task was executed or failed.
+
+
+.. image:: images/pipeline-step-2.png
```
Comment (kajayi) [inline papers/ido_michael/index.rst]: Needs a caption

File: papers/ido_michael/index.rst [hunk 20]
```diff
@@ -0,0 +1,403 @@
...
+is by ploting the pipeline (note that we’re now using ``ploomber``, which
+is the framework for developing pipelines):
+
+.. code:: sh
+
+   ploomber plot
+
+This command will generate the plot below for us, that way we can stay up to date
+with changes that are happening in our pipeline and get an immidate status of what
+task was executed or failed.
```
Comment (kajayi) [inline papers/ido_michael/index.rst]:
```suggestion
This command will generate the plot below for us, which will allow us to stay up to date
with changes that are happening in our pipeline and get the current status of
tasks that were executed or failed to execute.
```

File: papers/ido_michael/index.rst [hunk 21]
```diff
@@ -0,0 +1,403 @@
...
+We can see that we have a few new files. ``pipeline.yaml`` contains the
+pipeline declaration, and ``tasks/`` contains the *stages* that Soorgeon
+identified based on our H2 Markdown headings.
+
+.. code:: sh
+
+   ls playground/tasks
+
+One of the best ways to onboard new people and explain what each workflow is doing
+is by ploting the pipeline (note that we’re now using ``ploomber``, which
```
Comment (kajayi) [inline papers/ido_michael/index.rst]:
```suggestion
is by plotting the pipeline (note that we’re now using ``ploomber``, which
```

File: papers/ido_michael/index.rst [hunk 22]
```diff
@@ -0,0 +1,403 @@
...
+After running the refactor command, we can take a look at the local directory,
+and we'll see that now we have multiple python tasks which are ready for production:
+
+.. code:: sh
+
+   ls playground
+
+We can see that we have a few new files. ``pipeline.yaml`` contains the
+pipeline declaration, and ``tasks/`` contains the *stages* that Soorgeon
+identified based on our H2 Markdown headings.
```
Comment (kajayi) [inline papers/ido_michael/index.rst]:
```suggestion
identified based on our H2 Markdown headings:
```

File: papers/ido_michael/index.rst [hunk 23]
```diff
@@ -0,0 +1,403 @@
...
+it into smaller tasks the way they see fit.
+
+To refactor the notebook, we use the ``soorgeon refactor`` command:
+
+.. code:: sh
+
+   soorgeon refactor nb.ipynb
+
+After running the refactor command, we can take a look at the local directory,
+and we'll see that now we have multiple python tasks which are ready for production:
```
Comment (kajayi) [inline papers/ido_michael/index.rst]:
```suggestion
After running the refactor command, we can take a look at the local directory
and see that we now have multiple python tasks that are ready for production:
```

File: papers/ido_michael/index.rst [hunk 24]
```diff
@@ -0,0 +1,403 @@
...
+
+If you already have a python project in a single notebook, you can use our tool
+`Soorgeon <https://github.com/ploomber/soorgeon>`__ to automatically
+refactor it into a `Ploomber <https://github.com/ploomber/ploomber>`__
+pipeline. Soorgeon statically analyze your code, clean up unnecessary imports,
+and makes sure your monolithic notebook is broken down into smaller components.
+It does that by scanning the markdown in the notebook and analyzing the headers,
+each H2 header in our example is marking a new self-contained task. In addition,
+it can transform a notebook to a single-task pipeline and then the user can split
+it into smaller tasks the way they see fit.
```
Comment (kajayi) [inline papers/ido_michael/index.rst]:
```suggestion
it into smaller tasks as they see fit.
```

File: papers/ido_michael/index.rst [hunk 25]
```diff
@@ -0,0 +1,403 @@
...
+
+2. Refactoring a legacy notebook
+--------------------------------
+
+If you already have a python project in a single notebook, you can use our tool
+`Soorgeon <https://github.com/ploomber/soorgeon>`__ to automatically
+refactor it into a `Ploomber <https://github.com/ploomber/ploomber>`__
+pipeline. Soorgeon statically analyze your code, clean up unnecessary imports,
+and makes sure your monolithic notebook is broken down into smaller components.
+It does that by scanning the markdown in the notebook and analyzing the headers,
```
Comment (kajayi) [inline papers/ido_michael/index.rst]:
```suggestion
It does that by scanning the markdown in the notebook and analyzing the headers;
```

File: papers/ido_michael/index.rst [hunk 26]
```diff
@@ -0,0 +1,403 @@
...
+Users can easily break down their analysis into multiple notebooks and
+execute them all with a single command.
+
+2. Refactoring a legacy notebook
+--------------------------------
+
+If you already have a python project in a single notebook, you can use our tool
+`Soorgeon <https://github.com/ploomber/soorgeon>`__ to automatically
+refactor it into a `Ploomber <https://github.com/ploomber/ploomber>`__
+pipeline. Soorgeon statically analyze your code, clean up unnecessary imports,
```
Comment (kajayi) [inline papers/ido_michael/index.rst]:
```suggestion
pipeline. Soorgeon statically analyzes your code, cleans up unnecessary imports,
```

File: papers/ido_michael/index.rst [hunk 27]
```diff
@@ -0,0 +1,403 @@
...
+order. To ensure our notebooks run, we must continuously test them to
+catch these problems.
+
+A second notable problem is the size of notebooks: the more cells we
+have, the more difficult it is to debug since there are more variables
+and code involved.
+
+Software engineers typically break down projects into multiple steps and
+test continuously to prevent broken and unmaintainable code. However,
+applying these ideas for data analysis requires extra work: multiple
```
Comment (kajayi) [inline papers/ido_michael/index.rst]:
```suggestion
applying these ideas for data analysis requires extra work&mdash;multiple
```

File: papers/ido_michael/index.rst [hunk 28]
```diff
@@ -0,0 +1,403 @@
...
+1. Introduction
+---------------
+
+Notebooks are an excellent environment for data exploration: they allow
+us to write code interactively and get visual feedback, providing an
+unbeatable experience for understanding our data.
+
+However, this convenience comes at a cost: if we are not careful about
+adding and removing code cells, we may have an irreproducible notebook.
+Arbitrary execution order is a prevalent problem: A `recent
```
Comment (kajayi) [inline papers/ido_michael/index.rst]:
```suggestion
Arbitrary execution order is a prevalent problem: a `recent
```

File: papers/ido_michael/index.rst [hunk 29]
```diff
@@ -0,0 +1,403 @@
...
+There's a binder link in there as well, so you can launch it instantly.
+
+1. Introduction
+---------------
+
+Notebooks are an excellent environment for data exploration: they allow
+us to write code interactively and get visual feedback, providing an
+unbeatable experience for understanding our data.
+
+However, this convenience comes at a cost: if we are not careful about
```
Comment (kajayi) [inline papers/ido_michael/index.rst]:
```suggestion
However, this convenience comes at a cost&mdash; if we are not careful about
```

File: papers/ido_michael/index.rst [hunk 30]
```diff
@@ -0,0 +1,403 @@
...
+
+   gbr = GradientBoostingRegressor()
+   gbr.fit(X_train, y_train)
+
+   y_pred = gbr.predict(X_test)
+   sns.scatterplot(x=y_test, y=y_pred)
+
+
+
+7. Incremental builds
```
Comment (kajayi) [inline papers/ido_michael/index.rst]:
```suggestion
8. Incremental builds
```

File: papers/ido_michael/index.rst [hunk 31]
```diff
@@ -0,0 +1,403 @@
...
+.. image:: images/new-task.png
+
+
+You can see that Ploomber recognizes the new file, but it does not have
+any dependency, so let’s tell Ploomber that it should execute after
+``train-test-split``:
+
+
+Open ``playground/tasks/gradient-boosting-regressor.py`` as a notebook by right-clicking on it and then ``Open With`` -> ``Notebook``:
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (kajayi) [inline papers/ido_michael/index.rst]: Consider putting this into a separate Markdown cell

File: papers/ido_michael/index.rst [hunk 32]
```diff
@@ -0,0 +1,403 @@
...
+
+    # Sample tests after the notebook ran
+    # Check task test input exists
+    assert Path(upstream['train-test-split']['X_test']).exists()
+
+    # Check task train input exists
+    assert Path(upstream['train-test-split']['y_train']).exists()
+
+    # Validating output type
+    assert 'pkl' in upstream['train-test-split']['X_test']
```
Comment (kajayi) [inline papers/ido_michael/index.rst:211]: Consider putting this into a separate Markdown cell

File: papers/ido_michael/index.rst [hunk 33]
```diff
@@ -0,0 +1,403 @@
...
+
+    # Sample data quality checks after loading the raw data
+    # Check nulls
+    assert not df['HouseAge'].isnull().values.any()
+
+    # Check a specific range - no outliers
+    assert df['HouseAge'].between(0,100).any()
+
+    # Exact expected row count
+    assert len(df) == 11085
```
Comment (kajayi) [inline papers/ido_michael/index.rst:196]: Consider putting this into a separate Markdown cell

**Comment by kajayi** [comment]
Hello, I have reviewed the paper and it (initially) looks good overall.  I added a number of line comments to the PR. (I didn't realize they would show up in the conversation like this!)   This paper definitely will be of interest to the SciPy community. 

I went through the tutorial online and got a CalledProcessError on Step 9 ("Parallel execution + Ploomber cloud execution"). That step is not shown in the paper, so maybe you meant to remove it from the demo?  

I also tried to complete the tutorial locally, but got an error on step 5--that error is likely due to the way I handle environments on my local machine. (I use the nb_conda_kernels extension from the base environment and specify the desired conda environment as the kernel for each notebook.) I don't know if you have to do anything about that, but I thought you should know just FYI. 

Let me know if you have any questions!

File: papers/ido_michael/index.rst [hunk 34]
```diff
@@ -0,0 +1,403 @@
...
+5. Testing and quality checks
+-----------------------------
+
+** Open `tasks/train-test-split.py` as a notebook by right-clicking on it and then `Open With` -> `Notebook` and add the following code after the cell with `# noqa`:
+
+.. code:: python
+
+    # Sample data quality checks after loading the raw data
+    # Check nulls
+    assert not df['HouseAge'].isnull().values.any()
```
Comment (kajayi) [inline papers/ido_michael/index.rst:190]: Is "values" needed here (i.e., would `df['HouseAge'].isnull().any()` work)?

**Review by kajayi — CHANGES_REQUESTED** [review]
Added comments to index.rst

**Comment by deniederhut** [comment]
@scoobies mark needs more review

**Comment by deniederhut** [comment]
@scoobies remove @govarguz from reviewers

**Comment by scoobies** [comment]
@govarguz removed from the reviewers list!

**Comment by deniederhut** [comment]
@scoobies mark pending comment

File: papers/ido_michael/index.rst (hunk 34 above)
Comment (idomic) [reply papers/ido_michael/index.rst:190]: It's not, removed it.

**Comment by idomic** [comment]
Thanks for reviewing @kajayi fixed comments and added the missing section!

**Comment by deniederhut** [comment]
Hi @kajayi ! Do you feel this paper is now ready to be included in the proceedings?

**Comment by kajayi** [comment]
> Hi @kajayi ! Do you feel this paper is now ready to be included in the proceedings?

Hi @deniederhut !  There are still a few unresolved suggestions in the paper, but they are all minor, so I believe this paper is ready to be included in the proceedings whether the suggestions are resolved or not.  I will incorporate its teachings into my own work where I can--thank you for the tutorial @idomic !

**Comment by deniederhut** [comment]
> There are still a few unresolved suggestions in the paper

Sorry I missed this! @idomic would you mind responding to the remaining comments?

**Comment by deniederhut** [comment]
@idomic have you had a chance to work on the rest of the review comments?

**Comment by idomic** [comment]
@deniederhut I'll do it over this weekend. I probably missed those comments.

**Comment by mepa** [comment]
@scoobies check references

**Comment by scoobies** [comment]
Checking the BibTeX entries failed with the following error: 
```
Couldn't find bibliography entry
```

**Comment by mepa** [comment]
Hi @idomic, thanks for your contribution! Could you please check the rendered pdf and fix any lines that bleed into the next column by adding newlines at appropriate places in the code: https://procbuild.scipy.org/download/idomic-2022

@cbcunc, @deniederhut, @stargaser - this paper includes hypertext links in lieu of a references section with citations.

**Comment by deniederhut** [comment]
Thanks for the heads-up Meghann. It seems like the paper should at least be citing the Jupyter project -- maybe also scikit-learn?

**Comment by idomic** [comment]
@deniederhut fixed pending changes (it's all in the commit)

File: papers/ido_michael/index.rst [hunk 35]
```diff
@@ -0,0 +1,484 @@
...
+   from pathlib import Path
+   import pickle
+
+   import seaborn as sns
+   from sklearn.ensemble import GradientBoostingRegressor
+
+   y_train = pickle.loads(Path(upstream['train-test-split']['y_train']).read_bytes())
+   y_test = pickle.loads(Path(upstream['train-test-split']['y_test']).read_bytes())
+   X_test = pickle.loads(Path(upstream['train-test-split']['X_test']).read_bytes())
+   X_train = pickle.loads(Path(upstream['train-test-split']['X_train']).read_bytes())
```
Comment (mepa) [inline papers/ido_michael/index.rst]: This section is running off the page

Screen Shot 2022-07-04 at 11 26 15 PM

Something like the following _might_ work
```
   y_train = pickle.loads(
       Path(upstream['train-test-split']['y_train']).read_bytes())
   y_test = pickle.loads(
       Path(upstream['train-test-split']['y_test']).read_bytes())
   X_test = pickle.loads(
       Path(upstream['train-test-split']['X_test']).read_bytes())
   X_train = pickle.loads(
       Path(upstream['train-test-split']['X_train']).read_bytes())
```

File: papers/ido_michael/index.rst [hunk 36]
```diff
@@ -0,0 +1,484 @@
...
+
+.. figure:: images/new-task-attached.png
+   :alt: new-pipeline-task-attached
+
+   The new task is attached to the pipeline
+
+
+Ploomber now recognizes our dependency declaration!
+
+Open ``playground/tasks/gradient-boosting-regressor.py`` as a notebook by right-clicking on it and then ``Open With`` -> ``Notebook`` and add the following code:
```
Comment (mepa) [inline papers/ido_michael/index.rst]: Screen Shot 2022-07-04 at 11 30 21 PM

Maybe put ``playground/tasks/gradient-boosting-regressor.py`` on its own line so it doesn't run into the margin?

File: papers/ido_michael/index.rst [hunk 37]
```diff
@@ -0,0 +1,484 @@
...
+   :alt: new-task
+
+   Now we see an independent new task
+
+You can see that Ploomber recognizes the new file, but it does not have
+any dependency, so let’s tell Ploomber that it should execute after
+``train-test-split``:
+
+
+Open ``playground/tasks/gradient-boosting-regressor.py`` as a notebook by right-clicking on it and then ``Open With`` -> ``Notebook``:
```
Comment (mepa) [inline papers/ido_michael/index.rst]: Maybe put ``playground/tasks/gradient-boosting-regressor.py`` on its own line to avoid bleeding into the margin?

**Comment by mepa** [comment]
@scoobies build paper

**Comment by scoobies** [comment]
building paper...

**Comment by scoobies** [comment]
{"data"=>{"info"=>"Build for paper idomic-2022 scheduled.  Note that builds are only executed if the current build attempt is more than 5 minutes old."}, "status"=>"success"}

File: papers/ido_michael/index.rst (hunk 35 above)
Comment (mepa) [reply papers/ido_michael/index.rst]: Screen Shot 2022-07-07 at 12 35 35 AM

Let's please try this instead, to make it more compact while also not running off the page!
```
y_train = pickle.loads(Path(
    upstream['train-test-split']['y_train']).read_bytes())
y_test = pickle.loads(Path(
    upstream['train-test-split']['y_test']).read_bytes())
X_test = pickle.loads(Path(
    upstream['train-test-split']['X_test']).read_bytes())
X_train = pickle.loads(Path(
    upstream['train-test-split']['X_train']).read_bytes())
```

File: papers/ido_michael/index.rst (hunk 36 above)
Comment (mepa) [reply papers/ido_michael/index.rst]: Screen Shot 2022-07-07 at 12 38 24 AM

Ah, sorry, the newline didn't work like I hoped. Maybe the following will work better
```
Open

.. code:: sh

   playground/tasks/gradient-boosting-regressor.py

as a notebook by right-clicking on
```

File: papers/ido_michael/index.rst (hunk 37 above)
Comment (mepa) [reply papers/ido_michael/index.rst]: We can try the same approach here, too, if it works in the [other case](https://github.com/scipy-conference/scipy_proceedings/pull/688#discussion_r915477513).

**Comment by cbcunc** [comment]
@scoobies mark ready

**Comment by cbcunc** [comment]
@scoobies mark pending-comment

**Comment by cbcunc** [comment]
@scoobies mark pending comment

**Comment by mepa** [comment]
Hi @idomic, thanks for trying those changes! Could we please try the latest formatting changes I suggested? Also, I'm just guessing at what will work. It may be more efficient (quicker to get the paper accepting into the proceedings) if you can please do the following.

1. After committing your changes, build the output by commenting
```
@scoobies build paper
```
or by clicking the button next to your paper at https://procbuild.scipy.org/.

2. After the paper build, check the generated pdf at https://procbuild.scipy.org/download/idomic-2022 to ensure that the formatting looks good (e.g., no overflow into the margin or next column).

**Comment by idomic** [comment]
@mepa done, I think it looks better now!

**Comment by mepa** [comment]
@scoobies build paper

**Comment by scoobies** [comment]
building paper...

**Comment by scoobies** [comment]
{"data"=>{"info"=>"Build for paper idomic-2022 scheduled.  Note that builds are only executed if the current build attempt is more than 5 minutes old."}, "status"=>"success"}

**Comment by mepa** [comment]
Thanks, @idomic - that worked! There are other long urls in the paper that could be wrapped in code cells in the same way as you did for `playground/tasks/gradient-boosting-regressor.py`, if you feel like fiddling with it before the final proceedings is published around July 30th. Anyhow, I'll go ahead and mark this paper as ready to include in the proceedings. Thanks for the contribution and best wishes on tomorrow's Ploomber tutorial!

**Comment by mepa** [comment]
@scoobies mark ready

**Pull request merged**